pkg/utils/patch.go

```diff
@@ -97,3 +102,94 @@ func PatchVolumeGroupSnapshotContent(
 
 	return newGroupSnapshotContent, nil
 }
+
+// Remove one or more finalizers from an object
+// if finalizers is not empty, only the specified finalizers will be removed
+func PatchRemoveFinalizers(object metav1.Object, client clientset.Interface, finalizers ...string) (metav1.Object, error) {
+	data, err := PatchOpsBytesToRemoveFinalizers(object, finalizers...)
+	if err != nil {
+		return nil, err
+	}
+	switch object.(type) {
+	case *crdv1.VolumeSnapshot:
+		obj, err := client.SnapshotV1().VolumeSnapshots(object.GetNamespace()).Patch(context.TODO(), object.GetName(), types.JSONPatchType, data, metav1.PatchOptions{})
+		if len(obj.Finalizers) == 0 {
+			// to satisfy some tests that requires nil rather than []string{}
+			obj.Finalizers = nil
+		}
+		return obj, err
+	case *crdv1alpha1.VolumeGroupSnapshot:
+		obj, err := client.GroupsnapshotV1alpha1().VolumeGroupSnapshots(object.GetNamespace()).Patch(context.TODO(), object.GetName(), types.JSONPatchType, data, metav1.PatchOptions{})
+		if len(obj.Finalizers) == 0 {
+			// to satisfy some tests that requires nil rather than []string{}
+			obj.Finalizers = nil
+		}
+		return obj, err
+	default:
+		return nil, errors.New("PatchRemoveFinalizers: unsupported object type")
+	}
+}
+
+func PatchRemoveFinalizersCorev1(object metav1.Object, client kubernetes.Interface, finalizers ...string) (metav1.Object, error) {
+	data, err := PatchOpsBytesToRemoveFinalizers(object, finalizers...)
+	if err != nil {
+		return nil, err
+	}
+	obj, err := client.CoreV1().Pods(object.GetNamespace()).Patch(context.TODO(), object.GetName(), types.JSONPatchType, data, metav1.PatchOptions{})
+	if len(obj.Finalizers) == 0 {
+		// to satisfy some tests that requires nil rather than []string{}
+		obj.Finalizers = nil
+	}
+	return obj, err
+}
+
+func PatchOpsToRemoveFinalizers(object metav1.Object, finalizers ...string) []PatchOp {
+	patches := []PatchOp{}
+	if len(finalizers) == 0 {
+		return patches
+	}
+	// remove only the specified finalizers
+	// get index of all finalizers to be removed
+	// sort the indexes in descending order
+	// remove the finalizers in descending order
+	// this is to avoid the need to adjust the index after each removal
+	indexes := []int{}
+	for _, finalizer := range finalizers {
+		for i, f := range object.GetFinalizers() {
+			if f == finalizer {
+				indexes = append(indexes, i)
+			}
+		}
+	}
+	slices.Sort(indexes)
+	slices.Reverse(indexes)
+	for _, i := range indexes {
+		patches = append(patches, PatchOp{
+			Op:   "remove",
+			Path: "/metadata/finalizers/" + fmt.Sprint(i),
+		})
+	}
```

Comment:
This is racy. Consider another controller / user that added a finalizer to an object when this loop runs. Since the PATCH removes indices, it will remove a wrong item.

Is there a way how to remove a _value_ and not an index via `Patch()`? If not, then please stick to `Update()`

Reply:
I believe there is.
One is to remove all then add all back in one call. Let me try that.

Reply:
If this is still considered racy, then will have to wait for https://github.com/json-patch/json-patch2/issues/18 and will move back to update. lmk.

Reply:
It probably still is racy.. hmm. but at least it external-snapshotter won't hang. User could see the finalizer they added isn't there. but at least this won't have the "wrong index" issue.

Another would be to use update but test it can get out of the "out of date, please apply again"

Reply:
Used delete all + add which would not be stuck in [`the object has been modified; please apply your changes to the latest version and try again`](https://github.com/kubernetes-sigs/kubebuilder/blob/66c3572c9b4b6b918c129eaa5b27abe842c54c2a/docs/book/src/reference/good-practices.md?plain=1#L51)

Any other watchers should wait until the objects have finalizers external-snapshotter added before doing it's own patching/updating to avoid race condition removing "user" added labels.

https://github.com/kubernetes-csi/external-snapshotter/pull/1019/commits/2e56da847422c99bc5c5dc88a65e80cd74436fc5

Reply:
If a controller is running into "the object has been modified" errors, then we should find out why is that happening. IMO - using unilateral patch is kind of a band-aid (which is race prone and dangerous).

Reply:
I expect many more projects to add volumesnapshots to their watch and perform processing. If you insists patch is dangerous here, then there needs to be tests that ensures update fail calls are recoverable within few seconds, not over 10 minutes which is what we have been seeing.

https://github.com/vmware-tanzu/velero/issues/7356

Reply:
@jsafrane @gnufied see if you like https://github.com/kubernetes-csi/external-snapshotter/pull/1023 better.

Reply:
If I understand it correctly, the current version does not fix the race. It blindly removes all finalizers and add those that were known at the time the controller processed VolumeSnapshot / VolumeSnapshotContent.
It will again erase any finalizers added in parallel to the controller.

> While it is a bandaid, it is better than nothing

No. You are breaking finalizers of someone else. This is not a good behavior.

I think the whole fear of "the object has been modified" error is unjustified. It tells the controller that it has been working with stale data. The controller should check what has changed and try again, if it's still applicable. Or it may discover that the work is not needed any longer. In this case, the VolumeSnapshot / Content should be re-queued with exp. backoff.

> there needs to be tests that ensures update fail calls in external-snapshotter are recoverable within few seconds, not over 10 minutes which is what we have been seeing.

If that's true then this is the part that needs to be fixed. And not work around it using patch blindly everywhere. Can you reliably reproduce the issue, e.g. in an unit test? It should be easy to debug what's causing the delay then.

Reply:
Yes we've been hitting this issue everywhere in our CI and other velero users have been hitting it as well.

Reply:
```
		{
			name:             "2-4 - successful remove Snapshot finalizer after update conflict",
			initialSnapshots: newSnapshotArray("snap2-4", "snapuid2-4", "claim2-4", "", classSilver, "", &False, nil, nil, nil, false, true, nil),
			initialClaims:    newClaimArray("claim2-4", "pvc-uid2-4", "1Gi", "volume2-4", v1.ClaimBound, &classEmpty),
			test:             testRemoveSnapshotFinalizerAfterUpdateConflict,
			expectSuccess:    true,
			errors: []reactorError{
				{"update", "volumesnapshots", errors.NewConflict(crdv1.Resource("volumesnapshots"), "snap2-4", nil)},
			},
		},
```
Added this case here.
https://github.com/kubernetes-csi/external-snapshotter/pull/1023#pullrequestreview-1902552534

Reply:
> reliably reproduce the issue

50% of the time prior to https://github.com/kubernetes-csi/external-snapshotter/pull/876
After 876 it's much improved but added update calls after that OCP QE noted still cause issues sometimes.